lib/utils/genmap/genmap.go

```diff
@@ -0,0 +1,317 @@
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package genmap
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/gravitational/teleport/api/utils/retryutils"
+)
+
+// Config configures a generative map. The only required parameter is the generator itself.
+type Config[K comparable, V any] struct {
+	// Generator is the closure used to generate values for a given key.
+	Generator func(context.Context, K) (V, error)
+
+	// RegenInterval is the interval on which entries are regenerated.
+	RegenInterval time.Duration
+
+	// Jitter is the jitter applied to the regen interval.
+	Jitter retryutils.Jitter
+
+	// MaxFailures is the maximum number of times generation can fail
+	// in a row before we give up (note: giving up is temporary, the next
+	// call to Get will start generation up again for that key).
+	MaxFailures int
+}
+
+var (
+	// ErrGenMapClosed is returned by blocking operations called on a GenMap that has been closed.
+	ErrGenMapClosed = errors.New("generative map has been permanently closed")
+
+	// ErrEntryTerminated is returned in the case that a specific target entry was terminated while
+	// the caller is waiting on it to initialize.
+	ErrEntryTerminated = errors.New("generative map entry was concurrently terminated")
+)
+
+// GenMap is a mapping used to pre-generate and store values that we want to always have
+// available *immediately*, such as client tls configs. It serves a purpose similar to FnCache, but
+// with the notable distinction that it attempts to load/reload values *before* they
+// are needed, so that callers never need to wait. Once a key is accessed for the first
+// time it continues to be automatically regenerated until it is explicitly removed or the
+// provided generator function returns MaxFailures errors in a row. Note that for most usecases
+// the FnCache is probably a better choice. This helper is only really beneficial in cases where
+// a small set of very frequently accessed values exist *and* said values are costly to generate *and*
+// said values need to be regenerated very frequently, and possibly in response to events.
+type GenMap[K comparable, V any] struct {
+	cfg     Config[K, V]
+	rw      sync.RWMutex
+	entries map[K]*entry[V]
+	ctx     context.Context
+	cancel  context.CancelFunc
+}
+
+// New sets up a new GenMap based on the supplied configuration.
+func New[K comparable, V any](cfg Config[K, V]) *GenMap[K, V] {
+	if cfg.Generator == nil {
+		// no program that specifies a nil genetor is valid
+		panic("genmap.New called with nil Generator value")
+	}
+
+	if cfg.RegenInterval < 1 {
+		cfg.RegenInterval = time.Minute
+	}
+
+	if cfg.Jitter == nil {
+		cfg.Jitter = retryutils.NewSeventhJitter()
+	}
+
+	if cfg.MaxFailures < 1 {
+		cfg.MaxFailures = 3
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
`ctx` should optionally be the child of a parent context; even if we want absolute control of when the context gets canceled we can make it a `context.WithCancel(context.WithoutCancel(parentCtx))`.